java-checks/src/main/java/org/sonar/java/checks/InnerStaticClassesCheck.java

```diff
@@ -0,0 +1,131 @@
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.Deque;
+import java.util.LinkedList;
+
+@Rule(
+  key = "S2694",
+  name = "Inner classes which do not reference their owning classes should be \"static\"",
+  tags = {"performance"},
+  priority = Priority.MAJOR)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.MEMORY_EFFICIENCY)
+@SqaleConstantRemediation("15min")
+public class InnerStaticClassesCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private JavaFileScannerContext context;
+  private Deque<Symbol> outerClasses = new LinkedList<>();
+  private Deque<Boolean> atLeastOneReference = new LinkedList<>();
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitClass(ClassTree tree) {
+    Symbol.TypeSymbol symbol = tree.symbol();
+    if (!tree.is(Tree.Kind.CLASS)) {
+      return;
+    }
+    outerClasses.push(symbol);
+    atLeastOneReference.push(Boolean.FALSE);
+    super.visitClass(tree);
+    Boolean oneReference = atLeastOneReference.pop();
+    outerClasses.pop();
+    if (!outerClasses.isEmpty() && !isInnerInnerClass(outerClasses, symbol) && !symbol.isStatic() && !oneReference) {
+      String named = symbol.name().isEmpty() ? "named " : "";
+      context.addIssue(tree, this, "Make this a " + named + "\"static\" inner class.");
+    }
+  }
+
+  private boolean isInnerInnerClass(Deque<Symbol> outerClasses, Symbol.TypeSymbol symbol) {
+    if (!symbol.isStatic() && outerClasses.size() > 1) {
+      for (Symbol outerClass : outerClasses) {
+        if (outerClass.isStatic()) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  public void visitIdentifier(IdentifierTree tree) {
+    super.visitIdentifier(tree);
+    checkSymbol(tree.symbol());
+  }
+
+  @Override
+  public void visitNewClass(NewClassTree tree) {
+    super.visitNewClass(tree);
+    checkSymbol(tree.symbolType().symbol());
+  }
+
+  private void checkSymbol(Symbol symbol) {
+    if (!atLeastOneReference.isEmpty() && !atLeastOneReference.peek() && referenceInstance(symbol)) {
+      atLeastOneReference.pop();
+      atLeastOneReference.push(Boolean.TRUE);
+    }
+  }
+
+  private boolean referenceInstance(Symbol symbol) {
+    Symbol owner = symbol.owner();
+    return (owner != null && !owner.equals(outerClasses.peek())) && (symbol.isUnknown() || (!symbol.isStatic() && fromInstance(owner)));
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)